specification/dartLangSpec.tex

```diff
@@ -18117,35 +18118,76 @@ \subsection{Yield-Each}
   It is a dynamic type error if the class of $o$
   is not a subtype of \code{Stream<$T_f$>}.
 Otherwise
+\item
+  If the stream associated with $m$ has been cancelled,
+  then execution of $s$ completes by returning without a value.
+  \commentary{In this case, the \code{\YIELD*} operation does
+  not begin to listen to the stream $o$.}
+Otherwise
+\item
+  The $o$ stream is listened to by invoking \code{$v_o$.listen}
+  with system provided arguments,
+  where $v_o$ is a fresh variable referencing the stream $o$.
+  Let $u$ be the stream subscription returned by the call to
+  \code{$p$.listen} and let $v$ be a fresh variable bound to $u$.
+\item
+  If the stream associated with $m$ is paused, then $u$ is immediately
+  paused as if by invoking \code{$v$.pause} with no arguments.
```

Comment:
The default semantic fallback is that if some sub-part of the semantics of a construct throws, so does the construct itself *unless otherwise stated*. In this case, nothing otherwise is stated, so if `$v$.pause` throws, so does the `await for` body evaluation. (Which then does say what happens in that case - it cancels the subscription and rethrows).

Now, *should it* work like that? Probably yes.
If `pause` throws ... the stream probably isn't paused. Something is deeply wrong with that stream, so it's better to exit that loop immediately, and stop handling events from the stream entirely. (We should ignore any events after calling `cancel`).

It's possible to be obnoxious about the exiting. Say:
```dart
label: try {
  yield whatnot;
} finally {
  break label;
}
```
This will swallow any control flow trying to leave the `try` block and continue unabated.
We can't entirely control what happens after that, if the stream subscription acts maliciously.
(And that's why C# disallowed control flow in `finally` blocks!)